### PR TITLE
feat(es/minifier): Common sub expression in seq in condition

### PR DESCRIPTION
**Pull request by Austaras**
**Description:**

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

File: crates/swc_ecma_minifier/src/compress/optimize/conditionals.rs [hunk 1]
```diff
@@ -616,6 +616,81 @@ impl Optimizer<'_> {
...
             }
 
+            (Expr::Seq(left), Expr::Seq(right)) => {
+                let last = left.exprs.len().min(right.exprs.len()) - 1;
+
+                let mut idx = 0;
+
+                while idx <= last
+                    && left.exprs[last - idx].eq_ignore_span(&right.exprs[last - idx])
+                    && !matches!(&*left.exprs[last - idx], Expr::Yield(..) | Expr::Fn(..))
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/compress/optimize/conditionals.rs:626]: I think we need `Expr::Arrow`, too.

Comment (Austaras) [reply crates/swc_ecma_minifier/src/compress/optimize/conditionals.rs:626]: Why? I just copy this from L356, and I don't believe the original is necessary

Comment (kdy1) [reply crates/swc_ecma_minifier/src/compress/optimize/conditionals.rs:626]: Ah, I see. I just thought `Expr::Arrow `is required because there's `Expr::Fn`

Comment (Austaras) [reply crates/swc_ecma_minifier/src/compress/optimize/conditionals.rs:626]: Do you remember why the original restriction is written?

File: crates/swc_ecma_minifier/tests/benches-full/victory.js [hunk 2]
```diff
@@ -6526,7 +6526,7 @@
                 remove: function(node) {
                     node.N && (node.N.P = node.P), node.P && (node.P.N = node.N), node.N = node.P = null;
                     var sibling, next, red, parent = node.U, left = node.L, right = node.R;
-                    if (next = left ? right ? RedBlackFirst(right) : left : right, parent ? parent.L === node ? parent.L = next : parent.R = next : this._ = next, left && right ? (red = next.C, next.C = node.C, next.L = left, left.U = next, next !== right ? (parent = next.U, next.U = node.U, node = next.R, parent.L = node, next.R = right, right.U = next) : (next.U = parent, parent = next, node = next.R)) : (red = node.C, node = next), node && (node.U = parent), !red) {
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/benches-full/victory.js:6529]: I can't find `right.U = next` from other code

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

---


swc-bump:
 - swc_ecma_minifier

File: crates/swc_ecma_minifier/tests/benches-full/victory.js (hunk 2 above)
Comment (Austaras) [reply crates/swc_ecma_minifier/tests/benches-full/victory.js:6529]: Oh, I wrote something completely wrong...

**Review by kdy1 — APPROVED** [review]
Thank you!

---


swc-bump:
 - swc_ecma_minifier

File: crates/swc_ecma_minifier/src/compress/optimize/conditionals.rs [hunk 3]
```diff
@@ -616,6 +616,87 @@ impl Optimizer<'_> {
...
+
+                let mut idx = 0;
+
+                while idx < min_len
+                    && left.exprs[left_len - idx - 1]
+                        .eq_ignore_span(&right.exprs[right_len - idx - 1])
+                    && !matches!(
+                        &*left.exprs[left_len - idx - 1],
+                        Expr::Yield(..) | Expr::Fn(..)
+                    )
```
Comment (jridgewell) [inline crates/swc_ecma_minifier/src/compress/optimize/conditionals.rs:632]: Nit: we could trim both the start and the end of the sequences.

**Review by jridgewell — APPROVED** [review]

**Pull request merged**